File: Components/DevExpressVCL/Demos/ExpressEditors Library/Delphi/AlertWindowDemo/AlertWindowDemoMain.pas
unit AlertWindowDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, Menus, cxGraphics, cxControls, cxLookAndFeels,
  cxLookAndFeelPainters, ImgList, cxContainer, cxEdit, cxListBox, ComCtrls,
  ToolWin, dxAlertWindow, StdCtrls, cxGeometry, cxClasses, BaseForm,
  cxListView, ExtCtrls, ShellAPI, StrUtils;

type
  TAlertWindowDemoForm = class(TfmBaseForm)
    ilToolBar: TcxImageList;
    dxAlertWindowManager1: TdxAlertWindowManager;
    ilMessages: TcxImageList;
    pmAlertWindow: TPopupMenu;
    pmUnqueueMessage: TMenuItem;
    pmDeleteMessage: TMenuItem;
    pmCopyMessage: TMenuItem;
    miOptions: TMenuItem;
    lvContactList: TcxListView;
    bvSpacerLeft: TBevel;
    bvSpacerRight: TBevel;
    bvSpacerBottom: TBevel;
    tmGenerateMessage: TTimer;
    ilContactList_XP: TcxImageList;
    ilContactList_7: TcxImageList;
    procedure FormCreate(Sender: TObject);
    procedure pmUnqueueMessageClick(Sender: TObject);
    procedure pmDeleteMessageClick(Sender: TObject);
    procedure pmCopyMessageClick(Sender: TObject);
    procedure miOptionsAlertWindowClick(Sender: TObject);
    procedure tmGenerateMessageTimer(Sender: TObject);
    procedure lvContactListDblClick(Sender: TObject);
    procedure pmAlertWindowPopup(Sender: TObject);
    procedure dxAlertWindowManager1ButtonClick(Sender: TObject;
      AAlertWindow: TdxAlertWindow; AButtonIndex: Integer);
    procedure FormShow(Sender: TObject);
    procedure dxAlertWindowManager1Initialize(Sender: TObject;
      AAlertWindow: TdxAlertWindow);
    procedure FormDestroy(Sender: TObject);
  private
    FContactMessagesToGenerate: Integer;
    FContactMessageCounters: array of Integer;

    procedure NewMessage(AContactIndex: Integer);
    function FindWindowByCaption(const ACaption: string): TdxAlertWindow;
    function GetCurrentMessage(AAlertWindow: TdxAlertWindow): TdxAlertWindowMessage;
    procedure InitializeGenerateMessagesTimer(ACount: Integer; AInterval: Cardinal);
  end;

var
  AlertWindowDemoForm: TAlertWindowDemoForm;

implementation

uses
  dxCore, Math, Clipbrd, AlertWindowDemoOptions;

{$R *.dfm}

procedure TAlertWindowDemoForm.FormCreate(Sender: TObject);
begin
  lvContactList.SmallImages := ilContactList_XP;
  Randomize;
end;

function TAlertWindowDemoForm.FindWindowByCaption(const ACaption: string): TdxAlertWindow;

  function IsWindowForMessagesGroup(AAlertWindow: TdxAlertWindow): Boolean;
  begin
    Result := (AAlertWindow <> nil) and (AAlertWindow.Tag = 1) and
      (AAlertWindow.VisibilityTransition <> awvtHiding);
  end;

var
  I: Integer;
begin
  Result := nil;
  for I := 0 to dxAlertWindowManager1.Count - 1 do
  begin
    Result := dxAlertWindowManager1.Items[I];
    if IsWindowForMessagesGroup(Result) and (Result.MessageList[0].Caption = ACaption) then
      Break
    else
      Result := nil;
  end;
end;

procedure TAlertWindowDemoForm.InitializeGenerateMessagesTimer(
  ACount: Integer; AInterval: Cardinal);
begin
  FContactMessagesToGenerate := ACount;
  tmGenerateMessage.Interval := AInterval;
  tmGenerateMessage.Enabled := True;
end;

function TAlertWindowDemoForm.GetCurrentMessage(
  AAlertWindow: TdxAlertWindow): TdxAlertWindowMessage;
begin
  Result := AAlertWindow.MessageList[AAlertWindow.CurrentMessageIndex];
end;

procedure TAlertWindowDemoForm.miOptionsAlertWindowClick(Sender: TObject);
begin
  FormOptions.LoadWindowOptions(dxAlertWindowManager1);
  FormOptions.Show;
end;

procedure TAlertWindowDemoForm.NewMessage(AContactIndex: Integer);

  function FormatMessageText(AMessageNumber, AContactIndex: Integer): string;
  const
    FormatTextMessage = 'Message #%d from %s.' + #13#10 +
      'Note: Every new message from %s will be displayed in this window.';
  var
    ACaption: string;
  begin
    ACaption := lvContactList.Items[AContactIndex].Caption;
    Result := Format(FormatTextMessage, [AMessageNumber, ACaption, ACaption]);
  end;

  procedure ShowNewMessage(var AAlertWindow: TdxAlertWindow; const AMessageText: string; AContactIndex: Integer);
  var
    ACaption: string;
  begin
    ACaption := lvContactList.Items[AContactIndex].Caption;
    if AAlertWindow = nil then
    begin
      AAlertWIndow := dxAlertWindowManager1.Show(ACaption, AMessageText, AContactIndex);
      AAlertWIndow.Tag := 1;
    end
    else
    begin
      AAlertWindow.MessageList.Add(ACaption, AMessageText, AContactIndex);
      AAlertWindow.RestartDisplayTimer;
    end;
  end;

  procedure AddNavigationInfoToMessage(AAlertWindow: TdxAlertWindow; AIndexMessage: Integer);
  begin
    AAlertWindow.MessageList[AIndexMessage].Text := AAlertWindow.MessageList[AIndexMessage].Text
      + #13#10 + #13#10 + 'Use navigation buttons below to browse the message queue.';
  end;

var
  AMessageNumber: Integer;
  AAlertWindow: TdxAlertWindow;
begin
  if lvContactList.Items.Count >= 0 then
  begin
    AAlertWindow := FindWindowByCaption(lvContactList.Items[AContactIndex].Caption);
    Inc(FContactMessageCounters[AContactIndex]);
    AMessageNumber := FContactMessageCounters[AContactIndex];
    ShowNewMessage(AAlertWindow, FormatMessageText(AMessageNumber, AContactIndex), AContactIndex);
    if AAlertWindow.MessageList.Count > 1 then
    begin
      AddNavigationInfoToMessage(AAlertWindow, AAlertWindow.MessageList.Count - 1);
      if AAlertWindow.MessageList.Count = 2 then
        AddNavigationInfoToMessage(AAlertWindow, 0);
    end;
  end;
end;

procedure TAlertWindowDemoForm.pmCopyMessageClick(Sender: TObject);
begin
  Clipboard.AsText := GetCurrentMessage(pmAlertWindow.PopupComponent as TdxAlertWindow).Text;
end;

procedure TAlertWindowDemoForm.pmDeleteMessageClick(Sender: TObject);
begin
  (pmAlertWindow.PopupComponent as TdxAlertWindow).DeleteCurrentMessage;
end;

procedure TAlertWindowDemoForm.pmUnqueueMessageClick(Sender: TObject);
var
  AAlertWindow: TdxAlertWindow;
  AMessage: TdxAlertWindowMessage;
begin
  AAlertWindow := (pmAlertWindow.PopupComponent as TdxAlertWindow);
  AMessage := GetCurrentMessage(AAlertWindow);

  dxAlertWindowManager1.Show(AMessage.Caption,
    Copy(AMessage.Text, 1, Pos(#13#10, AMessage.Text) - 1), AMessage.ImageIndex);
  AAlertWindow.DeleteCurrentMessage;
end;

procedure TAlertWindowDemoForm.tmGenerateMessageTimer(Sender: TObject);
begin
  Dec(FContactMessagesToGenerate);
  NewMessage(RandomRange(0, lvContactList.Items.Count));
  if FContactMessagesToGenerate = 0 then
  begin
    tmGenerateMessage.Enabled := False;
    InitializeGenerateMessagesTimer(1, RandomRange(5000, 10000));
  end;
end;

procedure TAlertWindowDemoForm.lvContactListDblClick(Sender: TObject);
begin
  if lvContactList.ItemIndex > -1 then
    NewMessage(lvContactList.ItemIndex);
end;

procedure TAlertWindowDemoForm.pmAlertWindowPopup(Sender: TObject);
begin
  pmUnqueueMessage.Enabled := (pmAlertWindow.PopupComponent.Tag = 1) and
    ((pmAlertWindow.PopupComponent as TdxAlertWindow).MessageList.Count > 1);
end;

procedure TAlertWindowDemoForm.dxAlertWindowManager1ButtonClick(
  Sender: TObject; AAlertWindow: TdxAlertWindow; AButtonIndex: Integer);
var
  AAddressMessage: string;
begin
  if AButtonIndex = 0 then
  begin
    AAddressMessage := AnsiReplaceStr(AAlertWindow.MessageList[
      AAlertWindow.CurrentMessageIndex].Caption, ' ', '%20');
    ShellExecute(Handle, 'Open', PChar('mailto:' + AAddressMessage + '?'),
      nil, nil, sw_restore)
  end
  else
    if AButtonIndex = 1 then
      ShowMessage('Clicked!');
end;

procedure TAlertWindowDemoForm.FormShow(Sender: TObject);
begin
  SetLength(FContactMessageCounters, lvContactList.Items.Count);
  InitializeGenerateMessagesTimer(2, 1000);
end;

procedure TAlertWindowDemoForm.dxAlertWindowManager1Initialize(
  Sender: TObject; AAlertWindow: TdxAlertWindow);
begin
  AAlertWindow.MessageList[0].Text := AAlertWindow.MessageList[0].Text + #13#13 +
    'First message sent: ' + DateTimeToStr(Now);
end;

procedure TAlertWindowDemoForm.FormDestroy(Sender: TObject);
begin
//
end;

end.

File: Components/DevExpressVCL/Demos/ExpressEditors Library/CBuilder/AlertWindowDemo/AlertWindowDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "AlertWindowDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxListView"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "dxAlertWindow"
#pragma resource "*.dfm"
TAlertWindowDemoForm *AlertWindowDemoForm;
//---------------------------------------------------------------------------
__fastcall TAlertWindowDemoForm::TAlertWindowDemoForm(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TAlertWindowDemoForm::FormCreate(TObject *Sender)
{
  if (IsWinVistaOrLater)
	lvContactList->SmallImages = ilContactList_7;
  else
	lvContactList->SmallImages = ilContactList_XP;
  Randomize;
}
//---------------------------------------------------------------------------

TdxAlertWindow* TAlertWindowDemoForm::FindWindowByCaption(const String ACaption)
{
  int I;
  TdxAlertWindow *AAlertWindow;
  AAlertWindow = NULL;
  for (I = 0; I<= dxAlertWindowManager1->Count - 1; I++)
  {
	AAlertWindow = dxAlertWindowManager1->Items[I];
	if (IsWindowForMessagesGroup(AAlertWindow) && (AAlertWindow->MessageList->Items[0]->Caption == ACaption))
	  break;
	else
	  AAlertWindow = NULL;
  }
  return(AAlertWindow);
}
//---------------------------------------------------------------------------

String TAlertWindowDemoForm::FormatMessageText(int AMessageNumber, int AContactIndex)
{
  const String FormatTextMessage = "Message #%d from %s.\nNote: Every new message from %s will be displayed in this window. Use navigation buttons below to browse the message queue.";
  String ACaption;
  ACaption = lvContactList->Items->Item[AContactIndex]->Caption;
  TVarRec V[] = {AMessageNumber, ACaption, ACaption};
  return(Format(FormatTextMessage, V, 3));
}
//---------------------------------------------------------------------------

void TAlertWindowDemoForm::InitializeGenerateMessagesTimer(int ACount, Cardinal AInterval)
{
  FContactMessagesToGenerate = ACount;
  tmGenerateMessage->Interval = AInterval;
  tmGenerateMessage->Enabled = True;
}
//---------------------------------------------------------------------------

TdxAlertWindowMessage* TAlertWindowDemoForm::GetCurrentMessage(TdxAlertWindow* AAlertWindow)
{
  return(AAlertWindow->MessageList->Items[AAlertWindow->CurrentMessageIndex]);
}
//---------------------------------------------------------------------------

Boolean TAlertWindowDemoForm::IsWindowForMessagesGroup(TdxAlertWindow* AAlertWindow)
{
  return((AAlertWindow != NULL) && (AAlertWindow->Tag == 1) &&
	(AAlertWindow->VisibilityTransition != awvtHiding));
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::miOptionsAlertWindowClick(TObject *Sender)
{
  FormOptions->LoadWindowOptions(dxAlertWindowManager1);
  FormOptions->Show();
}
//---------------------------------------------------------------------------

void TAlertWindowDemoForm::ShowNewMessage(const String AMessageText, int AContactIndex)
{
  String ACaption;
  TdxAlertWindow* AAlertWindow;
  ACaption = lvContactList->Items->Item[AContactIndex]->Caption;
  AAlertWindow = FindWindowByCaption(ACaption);
  if (AAlertWindow == NULL)
	dxAlertWindowManager1->Show(ACaption, AMessageText, AContactIndex)->Tag = 1;
  else
  {
	AAlertWindow->MessageList->Add(ACaption, AMessageText, AContactIndex);
	AAlertWindow->RestartDisplayTimer();
  }
}
//---------------------------------------------------------------------------

void TAlertWindowDemoForm::NewMessage(int AContactIndex)
{
  if (lvContactList->Items->Count >= 0)
  {
	FContactMessageCounters[AContactIndex]++;
	ShowNewMessage(FormatMessageText(FContactMessageCounters[AContactIndex],
	  AContactIndex), AContactIndex);
  }
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::pmCopyMessageClick(TObject *Sender)
{
  Clipboard()->AsText = GetCurrentMessage((TdxAlertWindow*)pmAlertWindow->PopupComponent)->Text;
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::pmDeleteMessageClick(TObject *Sender)
{
  ((TdxAlertWindow*)pmAlertWindow->PopupComponent)->DeleteCurrentMessage();
}
//---------------------------------------------------------------------------


void __fastcall TAlertWindowDemoForm::tmGenerateMessageTimer(TObject *Sender)
{
  FContactMessagesToGenerate--;
  NewMessage(RandomRange(0, lvContactList->Items->Count));
  if (FContactMessagesToGenerate == 0)
  {
	tmGenerateMessage->Enabled = False;
	InitializeGenerateMessagesTimer(1, RandomRange(5000, 10000));
  }
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::pmAlertWindowPopup(TObject *Sender)
{
  pmUnqueueMessage->Enabled = (pmAlertWindow->PopupComponent->Tag == 1) &&
    (((TdxAlertWindow*)pmAlertWindow->PopupComponent)->MessageList->Count > 1);
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::lvContactListDblClick(TObject *Sender)
{
  if (lvContactList->ItemIndex > -1)
    NewMessage(lvContactList->ItemIndex);
}
//---------------------------------------------------------------------------


void __fastcall TAlertWindowDemoForm::pmUnqueueMessageClick(TObject *Sender)
{
  TdxAlertWindow* AAlertWindow;
  TdxAlertWindowMessage* AMessage;
  AnsiString S;
  AAlertWindow = (TdxAlertWindow*)pmAlertWindow->PopupComponent;
  AMessage = GetCurrentMessage(AAlertWindow);
  S = AMessage->Text;
  S = S.SubString(1, S.Pos("\n"));
  dxAlertWindowManager1->Show(AMessage->Caption, S, AMessage->ImageIndex);
  AAlertWindow->DeleteCurrentMessage();
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::dxAlertWindowManager1ButtonClick(TObject *Sender,
          TdxAlertWindow *AAlertWindow, int AButtonIndex)
{
  if (AButtonIndex == 0){
	AnsiString AMailTo = "mailto:" +
	  VarToStr(AAlertWindow->MessageList->Items[AAlertWindow->CurrentMessageIndex]->Caption);
	AMailTo = AnsiReplaceStr(AMailTo, " ", "%20");
	ShellExecute(Handle, "OPEN", AMailTo.c_str(), NULL, NULL, SW_SHOWMAXIMIZED);
  }
  else
	if (AButtonIndex == 1)
	  ShowMessage("Clicked!");
}
//---------------------------------------------------------------------------


void __fastcall TAlertWindowDemoForm::FormShow(TObject *Sender)
{
  FContactMessageCounters = new int[lvContactList->Items->Count];
  memset(FContactMessageCounters, 0, lvContactList->Items->Count * sizeof(int));
  InitializeGenerateMessagesTimer(2, 1000);
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::dxAlertWindowManager1Initialize(TObject *Sender,
		  TdxAlertWindow *AAlertWindow)
{
  const String FormatTextMessage = "%s\nFirst message sent: %s";
  TVarRec V[] = {AAlertWindow->MessageList->Items[0]->Text, DateTimeToStr(Now())};
  AAlertWindow->MessageList->Items[0]->Text = Format(FormatTextMessage, V, 2);
}
//---------------------------------------------------------------------------

void __fastcall TAlertWindowDemoForm::FormDestroy(TObject *Sender)
{
  delete(FContactMessageCounters);
}
//---------------------------------------------------------------------------

